examples/credential-registry/src/lib.rs

```diff
@@ -0,0 +1,1933 @@
+        let credential_entry = CredentialEntry {
+            holder_id:        credential_info.holder_id,
+            holder_revocable: credential_info.holder_revocable,
+            valid_from:       credential_info.valid_from,
+            valid_until:      credential_info.valid_until,
+            credential_data:  state_builder.new_box(CredentialData {
+                commitment:      credential_info.commitment.clone(),
+                credential_type: credential_info.credential_type.clone(),
+            }),
+            revocation_nonce: 0,
+            revoked:          false,
+        };
+        let res = self.credentials.insert(credential_id, credential_entry);
+        ensure!(res.is_none(), ContractError::CredentialAlreadyExists);
+        Ok(schema.clone())
+    }
+
+    fn revoke_credential(
+        &mut self,
+        now: Timestamp,
+        credential_id: CredentialID,
+    ) -> ContractResult<()> {
+        let mut credential =
+            self.credentials.get_mut(&credential_id).ok_or(ContractError::CredentialNotFound)?;
+        let status = credential.get_status(now);
+        ensure!(
+            status == CredentialStatus::Active || status == CredentialStatus::NotActivated,
+            ContractError::IncorrectStatusBeforeRevocation
+        );
+        credential.revoked = true;
+        Ok(())
+    }
+
+    fn restore_credential(
+        &mut self,
+        now: Timestamp,
+        credential_id: CredentialID,
+    ) -> ContractResult<()> {
+        let mut credential =
+            self.credentials.get_mut(&credential_id).ok_or(ContractError::CredentialNotFound)?;
+        let status = credential.get_status(now);
+        // It is expected that the credential was previously revoked.
+        // It can expire by the time of restoring; in this case, it is still marked as
+        // `revoked: false`, but the status will be `Expired`.
+        ensure!(status == CredentialStatus::Revoked, ContractError::IncorrectStatusBeforeRestoring);
+        credential.revoked = false;
+        Ok(())
+    }
+
+    fn register_issuer_key(&mut self, key_index: u16, pk: PublicKeyEd25519) -> ContractResult<()> {
+        let res = self.issuer_keys.insert(key_index, pk);
+        ensure!(res.is_none(), ContractError::KeyAlreadyExists);
+        Ok(())
+    }
+
+    fn remove_issuer_key(&mut self, key_index: u16) -> ContractResult<()> {
+        self.issuer_keys.remove(&key_index);
+        Ok(())
+    }
+
+    fn register_revocation_key(
+        &mut self,
+        key_index: u16,
+        pk: PublicKeyEd25519,
+    ) -> ContractResult<()> {
+        let res = self.revocation_keys.insert(key_index, (pk, 0));
+        ensure!(res.is_none(), ContractError::KeyAlreadyExists);
+        Ok(())
+    }
+
+    fn view_revocation_key(&self, key_index: u16) -> ContractResult<(PublicKeyEd25519, u64)> {
+        self.revocation_keys.get(&key_index).map(|x| *x).ok_or(ContractError::KeyNotFound)
+    }
```

Comment:
And then here we'd just have a single entrypoint that returns all the revocation keys.

Reply:
Resolved by #276